File: IteratorPattern/IteratorPattern/Waitress.cs
using IteratorPattern.Iterators;
using IteratorPattern.Menus;

namespace IteratorPattern;

public class Waitress
{
    public IMenu PancakeHouseMenu { get; set; }

    public IMenu DinerMenu { get; set; }

    public Waitress(IMenu pancakeHouseMenu, IMenu dinerMenu)
    {
        PancakeHouseMenu = pancakeHouseMenu;
        DinerMenu = dinerMenu;
    }

    public void PrintMenu()
    {
        IIterator pancakeHouseIterator = PancakeHouseMenu.CreateIterator();
        IIterator dinerIterator = DinerMenu.CreateIterator();

        Console.WriteLine("==> Breakfast");
        PrintMenu(pancakeHouseIterator);
        Console.WriteLine("==> Lunch");
        PrintMenu(dinerIterator);
    }

    private void PrintMenu(IIterator iterator)
    {
        while (iterator.HasNext())
        {
            var menuItem = iterator.Next() as MenuItem;
            Console.WriteLine("Name: " + menuItem?.Name);
            Console.WriteLine("Description: " + menuItem?.Description);
            Console.WriteLine("Price: " + menuItem?.Price);
        }
    }
}

File: IteratorPattern_Java/src/main/java/Waitress.java
import Menus.Menu;
import Menus.MenuItem;
import java.util.Iterator;

public class Waitress
{
    public Menu PancakeHouseMenu;

    public Menu DinerMenu;

    public Waitress(Menu pancakeHouseMenu, Menu dinerMenu)
    {
        PancakeHouseMenu = pancakeHouseMenu;
        DinerMenu = dinerMenu;
    }

    public void PrintMenu()
    {
        Iterator pancakeHouseIterator = PancakeHouseMenu.CreateIterator();
        Iterator dinerIterator = DinerMenu.CreateIterator();

        System.out.println("==> Breakfast");
        PrintMenu(pancakeHouseIterator);
        System.out.println("==> Lunch");
        PrintMenu(dinerIterator);
    }

    private void PrintMenu(Iterator iterator)
    {
        while (iterator.hasNext())
        {
            var menuItem = (MenuItem) iterator.next();
            System.out.println("Name: " + menuItem.Name);
            System.out.println("Description: " + menuItem.Description);
            System.out.println("Price: " + menuItem.Price);
        }
    }
}
